File: DraftRightLinux/draftright/ui/payment_status_banner.py
"""Compact live-status banner shown inside QR / bank-transfer dialogs.

Subscribes to a :class:`PaymentStatusStream` (polling
``/payment/status/:ref`` in a background thread) and updates a
single-line GTK widget with the current state.  On success it auto-
dismisses the parent dialog via the ``on_confirmed`` callback after
``auto_dismiss_seconds``.
"""

from __future__ import annotations

from typing import Callable, Optional

import gi
gi.require_version("Gtk", "4.0")
from gi.repository import GLib, Gtk

from draftright.models.payment import PaymentStatus, PaymentStatusUpdate
from draftright.services.payment_service import PaymentStatusStream


class PaymentStatusBanner(Gtk.Box):
    """Status banner wired to a :class:`PaymentStatusStream`."""

    def __init__(
        self,
        stream: Optional[PaymentStatusStream],
        on_confirmed: Callable[[], None],
        auto_dismiss_seconds: float = 2.0,
    ):
        super().__init__(orientation=Gtk.Orientation.HORIZONTAL, spacing=8)
        self._stream = stream
        self._on_confirmed = on_confirmed
        self._auto_dismiss_seconds = auto_dismiss_seconds
        self._dismiss_source: Optional[int] = None

        self.set_margin_top(6)
        self.set_margin_bottom(6)
        self.set_margin_start(10)
        self.set_margin_end(10)
        self.add_css_class("card")

        self._icon = Gtk.Label(label="⏳")
        self._text = Gtk.Label(label="Waiting for payment…", xalign=0)
        self._text.set_hexpand(True)
        self._text.set_wrap(True)
        self.append(self._icon)
        self.append(self._text)

        if stream is None:
            self.set_visible(False)
        else:
            self._apply_status(PaymentStatus.PENDING)
            stream.subscribe(self._on_update_from_thread)

    def cancel(self) -> None:
        """Stop polling; safe to call multiple times."""
        if self._stream is not None:
            self._stream.cancel()
            self._stream = None

    # ------------------------------------------------------------------
    # Background-thread callback → marshal to GTK main loop
    # ------------------------------------------------------------------

    def _on_update_from_thread(self, update: PaymentStatusUpdate) -> None:
        GLib.idle_add(self._on_update_on_main, update)

    def _on_update_on_main(self, update: PaymentStatusUpdate) -> bool:
        self._apply_status(update.status)
        if update.status.is_success() and self._dismiss_source is None:
            self._dismiss_source = GLib.timeout_add(
                int(self._auto_dismiss_seconds * 1000),
                self._fire_confirmed,
            )
        return False  # one-shot

    def _fire_confirmed(self) -> bool:
        self._dismiss_source = None
        try:
            self._on_confirmed()
        finally:
            self.cancel()
        return False

    # ------------------------------------------------------------------
    # Visual state
    # ------------------------------------------------------------------

    def _apply_status(self, status: PaymentStatus) -> None:
        if status in (PaymentStatus.PENDING, PaymentStatus.NOT_FOUND, PaymentStatus.UNKNOWN):
            self._icon.set_label("⏳")
            self._text.set_label("Waiting for payment…")
        elif status == PaymentStatus.COMPLETED:
            self._icon.set_label("✅")
            self._text.set_label("Payment confirmed!")
        elif status == PaymentStatus.FAILED:
            self._icon.set_label("❌")
            self._text.set_label("Payment failed. Please try again.")
        elif status == PaymentStatus.EXPIRED:
            self._icon.set_label("⌛")
            self._text.set_label(
                "Took too long to confirm. If you already paid, check Subscription in a minute."
            )
        elif status == PaymentStatus.REFUNDED:
            self._icon.set_label("↩")
            self._text.set_label("Refunded.")

File: DraftRightWindows/DraftRightWindows/Views/PaymentStatusBanner.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Threading;
using DraftRightWindows.Models;
using WpfTextBlock = System.Windows.Controls.TextBlock;

namespace DraftRightWindows.Views;

/// <summary>
/// Compact live-status banner (WPF, #160) used inside <see cref="QrCheckoutDialog"/>
/// and <see cref="BankTransferDialog"/>. Subscribes to the payment-status stream
/// and renders pending → confirmed/failed/expired, auto-closing the parent via
/// <c>onConfirmed</c> after <see cref="AutoDismissDelay"/> on success.
/// </summary>
public sealed class PaymentStatusBanner : UserControl, IObserver<PaymentStatusUpdate>
{
    public TimeSpan AutoDismissDelay { get; set; } = TimeSpan.FromSeconds(2);

    private readonly WpfTextBlock _icon;
    private readonly WpfTextBlock _text;
    private readonly Border _bar;
    private readonly DispatcherTimer _autoDismiss;
    private readonly Action _onConfirmed;
    private IDisposable? _subscription;

    public PaymentStatusBanner(IObservable<PaymentStatusUpdate>? stream, Action onConfirmed)
    {
        _onConfirmed = onConfirmed;

        _icon = new WpfTextBlock { FontSize = 15, VerticalAlignment = VerticalAlignment.Center, Margin = new Thickness(0, 0, 8, 0) };
        _text = new WpfTextBlock { FontWeight = FontWeights.Bold, Foreground = Brushes.White, TextWrapping = TextWrapping.Wrap, VerticalAlignment = VerticalAlignment.Center };
        var row = new StackPanel { Orientation = System.Windows.Controls.Orientation.Horizontal };
        row.Children.Add(_icon);
        row.Children.Add(_text);
        _bar = new Border { Padding = new Thickness(10, 6, 10, 6), Child = row };
        Content = _bar;

        _autoDismiss = new DispatcherTimer { Interval = AutoDismissDelay };
        _autoDismiss.Tick += (_, _) => { _autoDismiss.Stop(); _onConfirmed(); };

        if (stream is null)
        {
            Visibility = Visibility.Collapsed;
        }
        else
        {
            ApplyStatus(PaymentStatus.Pending);
            _subscription = stream.Subscribe(this);
            Unloaded += (_, _) => _subscription?.Dispose();
        }
    }

    public void OnNext(PaymentStatusUpdate u)
    {
        if (!Dispatcher.CheckAccess()) { try { Dispatcher.BeginInvoke(new Action(() => OnNext(u))); } catch { } return; }
        ApplyStatus(u.Status);
        if (u.Status.IsSuccess())
        {
            _autoDismiss.Stop();
            _autoDismiss.Interval = AutoDismissDelay <= TimeSpan.Zero ? TimeSpan.FromMilliseconds(50) : AutoDismissDelay;
            _autoDismiss.Start();
        }
    }

    public void OnCompleted() { }
    public void OnError(Exception error) { }

    // Status → colour + icon + copy. One place, mirrors the WinForms banner.
    private void ApplyStatus(PaymentStatus status)
    {
        (Color bg, string icon, string text) = status switch
        {
            PaymentStatus.Completed => (Color.FromRgb(22, 101, 52), "✅", "Payment confirmed!"),
            PaymentStatus.Failed    => (Color.FromRgb(153, 27, 27), "❌", "Payment failed. Please try again."),
            PaymentStatus.Expired   => (Color.FromRgb(154, 52, 18), "⌛", "Took too long. If you already paid, check Subscription in a minute."),
            PaymentStatus.Refunded  => (Color.FromRgb(75, 85, 99), "↩", "Refunded."),
            _                       => (Color.FromRgb(30, 64, 175), "⏳", "Waiting for payment…"),
        };
        _bar.Background = new SolidColorBrush(bg);
        _icon.Text = icon;
        _text.Text = text;
    }
}

/// <summary>Shared read-only field grid rendered by the QR and bank-transfer
/// dialogs (WPF, #160). One implementation so the two never drift.</summary>
internal static class BankInfoTable
{
    public static void Render(Panel host, BankInfo info)
    {
        Row(host, "Bank", info.BankName, copyable: false);
        Row(host, "Account #", info.AccountNumber, copyable: true);
        Row(host, "Name", info.AccountName, copyable: false);
        Row(host, "Amount", $"{AmountString(info.Amount)} {info.Currency}", copyable: true);
        Row(host, "Reference", info.Reference, copyable: true,
            hint: "Must include this in the transfer description.");
    }

    private static void Row(Panel host, string label, string value, bool copyable, string? hint = null)
    {
        var grid = new Grid { Margin = new Thickness(0, 0, 0, 6) };
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(90) });
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });

        var l = new WpfTextBlock { Text = label, Foreground = Theme.WpfBrush(Theme.TextMuted), VerticalAlignment = VerticalAlignment.Center };
        var box = new Wpf.Ui.Controls.TextBox
        {
            Text = value,
            IsReadOnly = true,
            FontFamily = new FontFamily("Cascadia Code, Consolas, monospace"),
            Margin = new Thickness(8, 0, 8, 0),
        };
        Grid.SetColumn(l, 0); Grid.SetColumn(box, 1);
        grid.Children.Add(l); grid.Children.Add(box);

        if (copyable)
        {
            var btn = FluentFormControls.SecondaryButton("Copy", () =>
            {
                try { System.Windows.Clipboard.SetText(value); } catch { /* clipboard busy */ }
            });
            btn.MinWidth = 64;
            Grid.SetColumn(btn, 2);
            grid.Children.Add(btn);
        }
        host.Children.Add(grid);

        if (!string.IsNullOrEmpty(hint))
        {
            host.Children.Add(new WpfTextBlock
            {
                Text = hint,
                FontStyle = FontStyles.Italic,
                FontSize = 12,
                Foreground = Theme.WpfBrush(Theme.TextMuted),
                Margin = new Thickness(98, 0, 0, 8),
                TextWrapping = TextWrapping.Wrap,
            });
        }
    }

    private static string AmountString(double v) =>
        v == Math.Floor(v) ? ((long)v).ToString() : v.ToString("F2");
}
